bolero-generator/src/arbitrary.rs

```diff
@@ -0,0 +1,70 @@
+use core::{cmp, marker::PhantomData};
+
+use arbitrary::{Arbitrary, Unstructured};
+
+use crate::{gen_with, Driver, ValueGenerator};
+
+pub struct ArbitraryGenerator<T>(PhantomData<T>);
+
+const ABUSIVE_SIZE: usize = 1024 * 1024;
+const MIN_INCREASE: usize = 32;
+
+impl<T> ValueGenerator for ArbitraryGenerator<T>
+where
+    T: for<'a> Arbitrary<'a>,
+{
+    type Output = T;
+
+    fn generate<D: Driver>(&self, driver: &mut D) -> Option<T> {
+        let size = T::size_hint(0);
+        let mut data = match T::size_hint(0) {
+            (min, Some(max)) if max < ABUSIVE_SIZE => gen_with::<Vec<u8>>()
+                .len(min..=max)
+                .generate(&mut *driver)?,
+            (min, _) => gen_with::<Vec<u8>>().len(min).generate(&mut *driver)?,
+        };
```

Comment:
While I appreciate not having to change anything, I'm thinking it might be more efficient to move this logic into the driver itself, since a lot of them are going to by byte-based anyway and we can potentially eliminate a copy. I'm thinking something like:

```rust
trait Driver {
    ...
    fn gen_from_bytes<F: FnOnce(&[u8]) -> Option<R>, R>(&mut self, len: Range<usize>, f: F) -> Option<R>;
}
```

I'm also wanting to support existing [proptest](https://altsysrq.github.io/rustdoc/proptest/1.0.0/proptest/test_runner/struct.TestRng.html) and [quickcheck](https://docs.rs/quickcheck/latest/quickcheck/trait.Arbitrary.html) generators so moving this into a shared location would make that easier.

Reply:
To be completely honest, I don't understand the `DriverMode` abstraction yet. I do understand that it's so that fuzzers can use `Direct` and cargo test can use `Forced`, but… why is it kept alongside the input for shrunk inputs, which indicates that byteslice drivers in forced mode do happen in certain circumstances? Why is it exposed from `FillBytes` and not from `Driver`? Why is there both `DirectRng` and `ForcedRng` despite the fact that a rng should have an infinite amount of data available anyway? (Ok for this one I actually know, it's because mode is also used for eg. ranged-int-generation) And, maybe even more than that, why is it even a property related to a `Driver`?

I'm thinking that semantically generating a value would require a `ValueGenerator` that shows how to generate a value, a `Driver` that gives a stream of data, and a `Mode` that says how to interpret invalid data. And right now, the `Mode` is merged inside the `Driver`.

But we could have instead:
- `ValueGenerator` as currently
- `Driver` generates a possibly-finite stream of data. `ExtendWithZeros(driver)` extends a driver with zeroes to emulate current `Forced` mode for byteslice and rng.
- `enum GenerationMode { FailIfInvalid, RepairIfInvalid }` is the current `DriverMode`, and is not a property of the driver nor is it a property of the generator.

Then, when generating a value, it just needs additionally setting the driver mode. `cargo test` could still default to `RepairIfInvalid & ExtendWithZeros` and `cargo bolero test` to `FailIfInvalid`.

Does that make sense?

Anyway, I've implemented what you suggested in this PR, but it ended up requiring quite a few other refactorings. WDYT? (I've also significantly reduced `ABUSIVE_SIZE`, so that tests pass quickly without generating megabytes worth of data for a `Vec`. Anyway, `arbitrary` and similar were designed to work well with fuzzers, that rarely generate more than 4-8k worth of data)

Reply:
Yeah there may be some conflation of the drivers and modes that could be simplified. But I don't want to conflate those changes with this PR. I think it looks great as is!